In [5]:
import pandas as pd
import gurobipy as gp
from gurobipy import GRB
import scipy.sparse as sp
import numpy as np

In [2]:
brews_df = pd.read_pickle('data/brews_with_costs.pkl')
ingredients_df = pd.read_pickle('data/ingredients.pkl')
brews_df

first_ingredient       second_ingredient     third_ingredient  \
0      Abecean Longfin      Ancestor Moth Wing  Cyrodilic Spadetail   
1      Abecean Longfin       Ash Creep Cluster         Beehive Husk   
2      Abecean Longfin       Ash Creep Cluster         Chaurus Eggs   
3      Abecean Longfin       Ash Creep Cluster  Cyrodilic Spadetail   
4      Abecean Longfin       Ash Creep Cluster            Elves Ear   
...                ...                     ...                  ...   
25237    Tundra Cotton              Void Salts                   NA   
25238    Tundra Cotton          Wisp Wrappings                   NA   
25239     Vampire Dust               White Cap                   NA   
25240       Void Salts          Wisp Wrappings                   NA   
25241            Wheat  Yellow Mountain Flower                   NA   

                                      effects                            name  \
0       [Fortify Restoration, Damage Stamina]   Potion of Fortify Restoration   
1        [Fortify Sneak, Fortify Destruction]   Potion of Fortify Destruction   
2          [Weakness to Poison, Invisibility]          Potion of Invisibility   
3       [Fortify Restoration, Damage Stamina]   Potion of Fortify Restoration   
4            [Weakness to Frost, Resist Fire]           Potion of Resist Fire   
...                                       ...                             ...   
25237         [Resist Magic, Fortify Magicka]       Potion of Fortify Magicka   
25238                          [Resist Magic]          Potion of Resist Magic   
25239                       [Restore Magicka]       Potion of Restore Magicka   
25240                          [Resist Magic]          Potion of Resist Magic   
25241  [Fortify Health, Damage Stamina Regen]  Poison of Damage Stamina Regen   

       value                                       descriptions  
0        243  [Drain the target's Stamina by 22 points., Res...  
1        432  [Destruction spells are 46% stronger for 60 se...  
2        486  [Invisibility for 37 seconds., Target is 15% w...  
3        243  [Drain the target's Stamina by 22 points., Res...  
4        190  [Resist 28% of fire damage for 60 seconds., Ta...  
...      ...                                                ...  
25237    194  [Magicka is increased by 37 points for 60 seco...  
25238     80               [Resist 9% of magic for 60 seconds.]  
25239     52                    [Restore 58 points of Magicka.]  
25240     80               [Resist 9% of magic for 60 seconds.]  
25241    356  [Decrease the target's Stamina regeneration by...  

[25242 rows x 7 columns]

In [28]:
brews = list(brews_df.name)
ingredients = list(ingredients_df.name)

The model is the following:

#### Parameters

$B = \{1, \dots, 25242\}$: Set of brews

$I = \{1, \dots, 109\}$: Set of ingredients

$I_p \subset{I}$: Set of ingredients required to make one unit of brew $b$ (note that $|I_b| \leq 3 \,\,\forall b \in B$)

$z_i \in \mathbb{Z}^+$: Amount of ingredient $i$ available

$c_b \in \mathbb{R}^+$: gold value of brew $b$

#### Decision Variables

$y_b \in \mathbb{Z}^+$: Number of brews $b$ created

$x_{ib} \in \mathbb{Z}^+$: Amount of ingredient $i$ used to create brew $b$

#### Model

$$ \max \sum_{i \in b} c_b y_p$$
such that:
$$y_b \leq x_{ib} \qquad \forall i \in I_b,  \forall b \in B$$
$$\sum_{\{b \mid i \in I_b\}} x_{ib} \leq z_i \qquad \forall i \in I$$


An additional sophistication for the model is to allow the purchase of common ingredients, which would amount to adding a integer decision variable for each such ingredient, adding them to the RHSs of the second constraints, and subtracting the costs in the objective.

In [6]:
B = [i for i in range(len(brews_df))]
I = [i for i in range(len(ingredients_df))]
c = np.array(brews_df['value'])

In [7]:
A = [[] for b in B]
for b in B:
    brew_ingredients = brews_df.loc[b][['first_ingredient','second_ingredient','third_ingredient']]
    for ingredient in brew_ingredients:
        if ingredient != 'NA':
            i = ingredients.index(ingredient)
            A[b].append(i)
            
# randomly generate inventory
z = np.random.poisson(lam=1, size=len(I))

Using Gurobi:

In [60]:
l = gp.tuplelist()
for b in B:
    for i in A[b]:
        l.append((b,i))
        
m = gp.Model('alchemy')
x = m.addVars(l, vtype=GRB.CONTINUOUS, name='x')
y = m.addVars(B, vtype=GRB.INTEGER, name='y')
m.addConstrs((y[b] <= x[b,i] for (b,i) in l), 'c1')
m.addConstrs((x.sum('*', i) <= z[i] for i in I))

m.setObjective(gp.quicksum(c[b]*y[b] for b in B), GRB.MAXIMIZE)
m.setParam('OUTPUT_FLAG',False)
m.optimize()

In [64]:
optimal_brews = [b for b in B if y[b].x > 0.5]
optimal_df = brews_df.loc[optimal_brews]
optimal_df['count'] = [int(y[b].x) for b in B if y[b].x > 0.5]
optimal_df.sort_values(by=['value'], ascending=False)
optimal_df = optimal_df[['name', 'count', 'value', 'first_ingredient', 'second_ingredient', 'third_ingredient', 'descriptions']]
optimal_df

name  count  value      first_ingredient  \
47          Potion of Fortify Restoration      1    380       Abecean Longfin   
693        Poison of Damage Magicka Regen      1    606    Ancestor Moth Wing   
1050       Poison of Damage Magicka Regen      1    589    Ancestor Moth Wing   
1194       Poison of Damage Magicka Regen      1    494    Ancestor Moth Wing   
1698               Potion of Invisibility      1    663     Ash Creep Cluster   
2618                  Poison of Paralysis      1    630       Ashen Grass Pod   
2840                  Poison of Paralysis      1    590       Ashen Grass Pod   
3006       Poison of Damage Magicka Regen      1    704            Bear Claws   
4755       Poison of Damage Magicka Regen      1    505        Bleeding Crown   
6070       Poison of Damage Stamina Regen      1    474         Blue Dartwing   
6852       Poison of Damage Magicka Regen      1    823  Blue Mountain Flower   
7126            Potion of Fortify Stamina      2   6588             Boar Tusk   
8623                       Poison of Slow      1   5034   Burnt Spriggan Wood   
8707       Poison of Damage Magicka Regen      1    820   Burnt Spriggan Wood   
9038                     Poison of Frenzy      1    253        Butterfly Wing   
11849      Potion of Fortify Carry Weight      1    727         Creep Cluster   
12368              Potion of Invisibility      1    785      Crimson Nirnroot   
12721                 Poison of Paralysis      1    632   Cyrodilic Spadetail   
13732        Potion of Regenerate Stamina      1    766       Dragon's Tongue   
14259        Potion of Regenerate Magicka      1    553           Dwarven Oil   
14687             Poison of Damage Health      1    681             Ectoplasm   
15073               Potion of Resist Fire      1    261             Elves Ear   
15183         Potion of Regenerate Health      1    804  Emperor Parasol Moss   
15456         Potion of Regenerate Health      1    590  Emperor Parasol Moss   
17494      Poison of Damage Stamina Regen      1    546         Frost Mirriam   
18351        Potion of Regenerate Magicka      1    588                Garlic   
19798             Potion of Resist Poison      1    323      Glowing Mushroom   
20415  Poison of Lingering Damage Magicka      1    192         Hagraven Claw   
21011              Potion of Resist Frost      1    332             Hawk Beak   
22011                 Poison of Paralysis      2    584           Human Flesh   
22083                 Poison of Paralysis      1    777           Human Flesh   
23114                      Poison of Slow      1   5232         Large Antlers   
24218        Potion of Fortify Pickpocket      1    458       Orange Dartwing   
25015            Poison of Ravage Stamina      1     62            Scathecraw   
25223      Poison of Damage Stamina Regen      1    468       Torchbug Thorax   

          second_ingredient        third_ingredient  \
47          Ashen Grass Pod     Cyrodilic Spadetail   
693     Blue Butterfly Wing                      NA   
1050      Hagraven Feathers              Spider Egg   
1194             Spider Egg                      NA   
1698       Glowing Mushroom            Vampire Dust   
2618            Briar Heart             Human Flesh   
2840            Human Flesh        Swamp Fungal Pod   
3006   Blue Mountain Flower        Rock Warbler Egg   
4755             Hawk's Egg              Nightshade   
6070           Daedra Heart            Skeever Tail   
6852              Glow Dust        Glowing Mushroom   
7126                 Garlic       Slaughterfish Egg   
8623          Large Antlers             River Betty   
8707              Salt Pile              Spider Egg   
9038      Hagraven Feathers               Troll Fat   
11849             Hawk Beak        Silverside Perch   
12368          Gleamblossom            Vampire Dust   
12721           Human Flesh             Netch Jelly   
13732           Fly Amanita          Mora Tapinella   
14259               

In [45]:
print(f"Optimal Objective Value: {int(m.objVal)}")

Optimal Objective Value: 41686
